### PR TITLE
CORDA-1217 Replace Guava caches with Caffeine

### PR DESCRIPTION
**Pull request by rick-r3**
Replace the buggy Guava caches with Caffeine.  https://github.com/ben-manes/caffeine

### CODE REVIEW

File: client/rpc/build.gradle [hunk 1]
```diff
@@ -68,6 +68,9 @@ dependencies {
     compile project(':core')
     compile project(':node-api')
 
+    // For caches rather than guava
+    compile "com.github.ben-manes.caffeine:caffeine:2.6.2"
```
Comment (rick-r3) [inline client/rpc/build.gradle]: Darn.  Think I should parameterise this.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClient.kt [hunk 2]
```diff
@@ -44,12 +44,6 @@ data class RPCClientConfiguration(
         val reapInterval: Duration,
         /** The number of threads to use for observations (for executing [Observable.onNext]) */
         val observationExecutorPoolSize: Int,
-        /**
```
Comment (rick-r3) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClient.kt]: Concurrency level is not configurable anymore.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 3]
```diff
@@ -154,10 +155,9 @@ class RPCClientProxyHandler(
             }
             observablesToReap.locked { observables.add(observableId) }
         }
-        return CacheBuilder.newBuilder().
+        return Caffeine.newBuilder().
                 weakValues().
-                removalListener(onObservableRemove).
-                concurrencyLevel(rpcConfiguration.cacheConcurrencyLevel).
+                removalListener(onObservableRemove).executor(SameThreadExecutor.getExecutor()).
```
Comment (exFalso) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: I don't think we need the samethreadexecutor trick for RPC

File: node/src/main/kotlin/net/corda/node/services/messaging/RPCServer.kt [hunk 4]
```diff
@@ -145,11 +146,11 @@ class RPCServer(
...
     private fun createObservableSubscriptionMap(): ObservableSubscriptionMap {
-        val onObservableRemove = RemovalListener<InvocationId, ObservableSubscription> {
-            log.debug { "Unsubscribing from Observable with id ${it.key} because of ${it.cause}" }
-            it.value.subscription.unsubscribe()
+        val onObservableRemove = RemovalListener<InvocationId, ObservableSubscription> { key, value, cause ->
+            log.debug { "Unsubscribing from Observable with id ${key} because of ${cause}" }
+            value!!.subscription.unsubscribe()
         }
-        return CacheBuilder.newBuilder().removalListener(onObservableRemove).build()
+        return Caffeine.newBuilder().removalListener(onObservableRemove).executor(SameThreadExecutor.getExecutor()).build()
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/messaging/RPCServer.kt]: Same, no need for the executor

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/messaging/RPCServer.kt]: I think we should not introduce any additional threads (it used ForkJoinPool) though which the default would do (and this maintains prior behaviour)

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 3 above)
Comment (rick-r3) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: I think we should not introduce any additional threads (it used ForkJoinPool) though which the default would do (and this maintains prior behaviour)

Comment (andr3ej) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: agreed, modifying the prior thread behaviour invites bugs

**Pull request merged**